src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -60,36 +88,76 @@ private void OnSolutionCrawlerProgressChanged(object sender, ProgressData progre
                 case ProgressStatus.Started:
                     StartedOrStopped(started: true);
                     break;
-                case ProgressStatus.Updated:
-                    ProgressUpdated(progressData.FilePathOpt);
+                case ProgressStatus.PendingItemCountUpdated:
+                    _lastPendingItemCount = progressData.PendingItemCount.Value;
+                    ProgressUpdated();
                     break;
                 case ProgressStatus.Stoped:
                     StartedOrStopped(started: false);
                     break;
+                case ProgressStatus.Evaluating:
+                case ProgressStatus.Paused:
+                    _lastProgressStatus = progressData.Status;
+                    ProgressUpdated();
+                    break;
                 default:
                     throw ExceptionUtilities.UnexpectedValue(progressData.Status);
             }
         }
 
-        private void ProgressUpdated(string filePathOpt)
+        private void ProgressUpdated()
         {
+            // we prefer showing evaluating if progress is flipping between evaluate and pause
+            // in short period of time.
+            var forceUpdate = _lastShownProgressStatus == ProgressStatus.Paused &&
+                              _lastProgressStatus == ProgressStatus.Evaluating;
+
             var current = DateTimeOffset.UtcNow;
-            if (current - _lastTimeReported < s_minimumInterval)
+            if (!forceUpdate && current - _lastTimeReported < s_minimumInterval)
             {
                 // make sure we are not flooding UI. 
-                // this is just presentation, fine to not updating UI especially since
+                // this is just presentation, fine to not updating UI right away especially since
                 // at the end, this notification will go away automatically
+                // but to make UI to be updated to right status eventually if task takes long time to finish
+                // we enqueue refresh task.
+                EnqueueRefresh();
                 return;
             }
 
+            _lastShownProgressStatus = _lastProgressStatus;
             _lastTimeReported = current;
-            ChangeProgress(_taskHandler, filePathOpt != null ? string.Format(ServicesVSResources.Analyzing_0, FileNameUtilities.GetFileName(filePathOpt)) : null);
+
+            ChangeProgress(_taskHandler, GetMessage());
+
+            string GetMessage()
+            {
+                var statusMessage = (_lastProgressStatus == ProgressStatus.Paused) ? ServicesVSResources.Paused_0_tasks_in_queue : ServicesVSResources.Evaluating_0_tasks_in_queue;
+                return string.Format(statusMessage, _lastPendingItemCount);
+            }
+
+            void EnqueueRefresh()
+            {
+                if (_resettableDelay != null)
+                {
+                    _resettableDelay.Reset();
```

Comment:
💡 Consider removing this call to `Reset`. There is a discrepancy between the interval check mentioned in https://github.com/dotnet/roslyn/pull/37943/files#r313468406 and this reset operation. If a long sequence of queueing operations occurs, the UI will never update with the current approach because each queue update will push the reset further into the future. For sequences like this, we still want to periodically update the UI.

Reply:
that is not how resettable delay is being used here. see this comment 
https://github.com/dotnet/roslyn/pull/37943/files#diff-fecb712d7896a08890d74844af27cb4cR32

Reply:
@heejaechang This may be the intent, but the implementation does not match that behavior. There are red flags with a high probability of bugs in the current implementation, based on the last two comments I made.